File: android/src/main/kotlin/io/sentient/android/settings/secrets/SecretsViewModel.kt
// ---------------------------------------------------------------------------
// SecretsViewModel — thin state-holder for the Secrets (admin) settings page.
// Presence-only key management over SettingsComponent.admin (AdminUseCases): GET
// returns has_key booleans (NEVER key material), PUT is write-only. Key values live
// ONLY in the SecretRow composable's local draft — they never enter this VM's state
// or any log line (logging carries provider name + a hasKey boolean, that's all).
//
// Restart affordance (webui parity): a saved key/base-url/active-change eagerly PUTs
// and flags `restartNeeded`. Webui queues that as a "slow" op on its apply-bar; the
// mobile mirror shows an inline "Restart to pick up the new key" notice + an
// "Apply now" button that fires the bare apply via ApplyProfileChangeUseCase.applyOnly.
// ---------------------------------------------------------------------------
package io.sentient.android.settings.secrets

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.sentient.mobiledata.result.SentientResult
import io.sentient.mobiledata.usecase.settings.AdminUseCases
import io.sentient.mobiledata.usecase.settings.ApplyProfileChangeUseCase
import io.sentient.mobiledata.usecase.settings.ApplyState
import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.settings.SecretsStatus
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** LLM provider slugs — must match the gateway secrets routes (secrets.ts LLM_PROVIDERS). */
object SecretProviders {
    const val OPENROUTER = "openrouter"
    const val OLLAMA_CLOUD = "ollama-cloud"
    const val CUSTOM = "custom"
}

/** Which row's inline field is open. */
enum class SecretField { KEY, BASE_URL }

data class EditingTarget(val provider: String, val field: SecretField)

data class SecretsUiState(
    val status: SecretsStatus? = null,
    val editing: EditingTarget? = null,
    val restartNeeded: Boolean = false,
    /** Non-null while / after a bare apply — drives the "Applying…/Applied/Failed" affordance. */
    val applying: ApplyState? = null,
    val errorMessage: String? = null,
)

class SecretsViewModel(
    private val admin: AdminUseCases,
    private val applyProfileChange: ApplyProfileChangeUseCase,
) : ViewModel() {
    private val log = createLogger("android", "settings", "secrets-vm")

    private val _state = MutableStateFlow(SecretsUiState())
    val state: StateFlow<SecretsUiState> = _state.asStateFlow()

    init {
        refresh()
    }

    fun refresh() {
        viewModelScope.launch {
            when (val r = admin.getSecretsStatus()) {
                is SentientResult.Success -> _state.update { it.copy(status = r.data) }
                is SentientResult.Failure -> _state.update { it.copy(errorMessage = "Failed to load provider keys.") }
                is SentientResult.Loading -> Unit
            }
        }
    }

    fun startEdit(provider: String, field: SecretField) =
        _state.update { it.copy(editing = EditingTarget(provider, field), errorMessage = null) }

    fun cancelEdit() = _state.update { it.copy(editing = null) }

    /** Save a provider key (write-only; value never logged/stored in state), then refetch. */
    fun saveKey(provider: String, value: String) = mutate("key.save", provider) {
        admin.setLlmProviderKey(provider, value = value, baseUrl = null)
    }

    /** Save the custom provider base URL, then refetch. */
    fun saveBaseUrl(url: String) = mutate("baseurl.save", SecretProviders.CUSTOM) {
        admin.setLlmProviderKey(SecretProviders.CUSTOM, value = null, baseUrl = url)
    }

    /** Point the active LLM provider at [provider], then refetch. */
    fun setActive(provider: String) = mutate("set-active", provider) {
        admin.setActiveLlmProvider(provider)
    }

    private inline fun mutate(op: String, provider: String, crossinline call: suspend () -> SentientResult<Unit>) {
        viewModelScope.launch {
            when (call()) {
                is SentientResult.Success -> {
                    log.info(op, mapOf("provider" to provider)) // value/url NEVER logged
                    _state.update { it.copy(editing = null, restartNeeded = true) }
                    refresh()
                }
                is SentientResult.Failure ->
                    _state.update { it.copy(errorMessage = "Couldn't save. Try again.") }
                is SentientResult.Loading -> Unit
            }
        }
    }

    /** Fire the bare apply so the worker restarts and picks up the new key. */
    fun applyNow() {
        if (_state.value.applying is ApplyState.Restarting) return
        viewModelScope.launch {
            applyProfileChange.applyOnly().collect { st ->
                _state.update {
                    it.copy(applying = st, restartNeeded = if (st is ApplyState.Ready) false else it.restartNeeded)
                }
            }
        }
    }

    /** Dismiss a terminal apply affordance (Applied/Already/Failed). */
    fun clearApplyResult() = _state.update { it.copy(applying = null) }
}

File: ios/App/Settings/Secrets/SecretsViewModel.swift
// ---------------------------------------------------------------------------
// SecretsViewModel — Admin "Secrets" page over `settings.admin` (provider keys)
// + `settings.applyProfileChange` (bare apply). Presence-only status (never key
// material), per-provider key update, active-provider selection, and the Custom
// base-URL. IMPERATIVE ops: PUT then refetch. After any successful key/base-URL/
// active change the running worker still holds the OLD key, so the page raises a
// "Restart assistant to pick up the new key" notice with an Apply-now button that
// drives `ApplyProfileChangeUseCase.applyOnly()` — the bare-apply entry the
// mobile-data layer settled on (POST /profile/apply, no mutation).
//
// Keys are NEVER logged (only provider name). @MainActor @Observable.
// ---------------------------------------------------------------------------
import Foundation
import MobileData

private let appliedResetDelayNs: UInt64 = 3_000_000_000

@MainActor
@Observable
final class SecretsViewModel {
    /// The three LLM providers; rawValue is the exact gateway path/key token.
    enum Provider: String, CaseIterable {
        case openrouter
        case ollamaCloud = "ollama-cloud"
        case custom
    }

    /// Which row (if any) is in edit mode. Only one at a time (webui parity).
    enum EditTarget: Equatable {
        case none
        case key(Provider)
        case customBaseUrl
    }

    /// The restart-notice / apply lifecycle.
    enum ApplyPhase: Equatable {
        case hidden
        case notice
        case applying
        case applied
        case alreadyApplying
        case failed(String)
    }

    private(set) var status: SecretsStatus?
    private(set) var isError = false
    private(set) var editing: EditTarget = .none
    private(set) var isSavingKey = false
    private(set) var apply: ApplyPhase = .hidden

    private let admin: AdminUseCases
    private let applyProfileChange: ApplyProfileChangeUseCase
    private let log = AppLog("settings", "secrets-vm")
    private var appliedResetTask: Task<Void, Never>?

    init(admin: AdminUseCases, applyProfileChange: ApplyProfileChangeUseCase) {
        self.admin = admin
        self.applyProfileChange = applyProfileChange
    }

    /// Load presence status. Idempotent; safe on each `.task`.
    func load() async {
        do {
            let result = try await admin.getSecretsStatus()
            switch onEnum(of: result) {
            case .success(let s):
                status = s.data
                isError = false
                log.info("secrets.loaded active=\(s.data.llm.active)")
            case .failure(let f):
                isError = true
                log.warn("secrets.load.failed kind=\(f.error.kind.name)")
            case .loading:
                break
            }
        } catch is CancellationError {
        } catch {
            isError = true
        }
    }

    func startEditKey(_ provider: Provider) { editing = .key(provider) }
    func startEditBaseUrl() { editing = .customBaseUrl }
    func cancelEdit() { editing = .none }

    /// Save a provider key (value never logged), refetch, then raise the restart notice.
    func saveKey(_ provider: Provider, value: String) async {
        await mutate("key.saved provider=\(provider.rawValue)") {
            try await self.admin.setLlmProviderKey(provider: provider.rawValue, value: value, baseUrl: nil)
        }
    }

    /// Save the Custom base URL, refetch, then raise the restart notice.
    func saveBaseUrl(_ url: String) async {
        await mutate("baseurl.saved") {
            try await self.admin.setLlmProviderKey(provider: Provider.custom.rawValue, value: nil, baseUrl: url)
        }
    }

    /// Select the active provider, refetch, then raise the restart notice.
    func setActive(_ provider: Provider) async {
        await mutate("active.set provider=\(provider.rawValue)") {
            try await self.admin.setActiveLlmProvider(provider: provider.rawValue)
        }
    }

    /// Apply now: restart the worker so the new key is picked up. Drives the FSM.
    func applyNow() async {
        apply = .applying
        for await state in applyProfileChange.applyOnly() {
            switch onEnum(of: state) {
            case .saving, .restarting:
                apply = .applying
            case .ready:
                apply = .applied
                log.info("apply.ready")
                scheduleAppliedReset()
            case .alreadyApplying:
                apply = .alreadyApplying
                log.warn("apply.already-applying")
            case .failed(let f):
                apply = .failed(f.error.userMessage)
                log.warn("apply.failed kind=\(f.error.kind.name)")
            case .idle:
                break
            }
        }
    }

    func dismissNotice() {
        appliedResetTask?.cancel()
        apply = .hidden
    }

    /// Run a secrets mutation, then refetch + raise the restart notice on success.
    private func mutate<T>(_ label: String, _ call: @escaping () async throws -> SentientResult<T>) async {
        isSavingKey = true
        defer { isSavingKey = false }
        do {
            let result = try await call()
            switch onEnum(of: result) {
            case .success:
                editing = .none
                log.info(label)
                await load()
                apply = .notice
            case .failure(let f):
                log.warn("secrets.mutate.failed kind=\(f.error.kind.name)")
            case .loading:
                break
            }
        } catch is CancellationError {
        } catch {
            log.warn("secrets.mutate.threw")
        }
    }

    private func scheduleAppliedReset() {
        appliedResetTask?.cancel()
        appliedResetTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: appliedResetDelayNs)
            guard let self, self.apply == .applied else { return }
            self.apply = .hidden
        }
    }
}
